client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,406 @@
+				let mut obj = serde_json::Map::new();
+				obj.insert("id".to_string(), id.into_u64().into());
+				obj.insert("payload".to_string(), connection_message.clone().into());
+				let (connection_messages, connection_notifiers) = node_args
+					.entry(addr.clone())
+					.or_insert_with(|| (Vec::new(), Vec::new()));
+				connection_messages.push(obj);
+				connection_notifiers.extend(connection_notifier_senders.clone());
+			}
+		}
 
-/// Implements `slog::Drain`.
-struct TelemetryDrain {
-	/// Sends log entries.
-	sender: std::panic::AssertUnwindSafe<mpsc::Sender<async_record::AsyncRecord>>,
-}
+		let mut node_pool: HashMap<Multiaddr, _> = existing_nodes
+			.iter()
+			.map(|addr| {
+				let (connection_messages, connection_notifiers) = node_args
+					.remove(addr)
+					.expect("there is a node for every connection message; qed");
+				let node = Node::new(
+					transport.clone(),
+					addr.clone(),
+					connection_messages,
+					connection_notifiers,
+				);
+				(addr.clone(), node)
+			})
+			.collect();
 
-/// Initializes the telemetry. See the crate root documentation for more information.
-///
-/// Please be careful to not call this function twice in the same program. The `slog` crate
-/// doesn't provide any way of knowing whether a global logger has already been registered.
-pub fn init_telemetry(config: TelemetryConfig) -> Telemetry {
-	// Build the list of telemetry endpoints.
-	let (endpoints, wasm_external_transport) = (config.endpoints.0, config.wasm_external_transport);
-
-	let (sender, receiver) = mpsc::channel(16);
-	let guard = {
-		let logger = TelemetryDrain { sender: std::panic::AssertUnwindSafe(sender) };
-		let root = slog::Logger::root(slog::Drain::fuse(logger), slog::o!());
-		slog_scope::set_global_logger(root)
-	};
-
-	let worker = match worker::TelemetryWorker::new(endpoints, wasm_external_transport) {
-		Ok(w) => Some(w),
-		Err(err) => {
-			error!(target: "telemetry", "Failed to initialize telemetry worker: {:?}", err);
-			None
+		let mut stream = receiver
+			.filter_map(|(id, verbosity, message): (Id, u8, String)| {
+				if let Some(nodes) = node_map.get(&id) {
+					future::ready(Some((verbosity, message, nodes)))
+				} else {
+					log::error!(
+						target: "telemetry",
+						"Received telemetry log for unknown id ({:?}): {}",
+						id,
+						message,
+					);
+					future::ready(None)
+				}
+			})
+			.flat_map(
+				|(verbosity, message, nodes): (u8, String, &Vec<(u8, Multiaddr)>)| {
+					let mut to_send = Vec::with_capacity(nodes.len());
+					let before = Instant::now();
+
+					for (node_max_verbosity, addr) in nodes {
+						if verbosity > *node_max_verbosity {
+							log::trace!(
+								target: "telemetry",
+								"Skipping {} for log entry with verbosity {:?}",
+								addr,
+								verbosity,
+							);
+							continue;
+						}
+
+						to_send.push((addr.clone(), message.clone()));
+					}
+
+					if before.elapsed() > Duration::from_millis(200) {
```

Comment:
This code is probably never taking more than 200ms. In the old code this was measuring the time it took to send a message.

Reply:
Yeah :thinking: doesnt make much sense anymore. Removed in 36791405d17da654fea1b7fe35568cb9ac43ee7a